opte/src/engine/packet.rs

```diff
@@ -751,19 +751,98 @@ impl Packet<Initialized> {
+        // inclusive upper bound of the squash.
+        let squash_to = match (body.seg_index, body.seg_offset) {
+            // The body is in the first segment meaning all headers are also in
+            // the first segment. No squashing needed.
+            (0, _) => 0,
+
+            // The body starts at a zero offset in segment n. This means we need
+            // to squash all segments prior to n.
+            (n, 0) => n - 1,
+
+            // The body starts at a non-zero offset in segment n. This means we
+            // need to squash all segments up to and including n.
+            (n, _) => n,
+        };
+
+        // If the squash bound is zero, there is notthing left to do here, just
+        // return.
+        if squash_to == 0 {
+            return Ok(Packet {
+                avail: self.avail,
+                source: self.source,
+                // The new packet is taking ownership of the segments.
+                segs: core::mem::take(&mut self.segs),
+                state: Parsed {
+                    len: self.state.len,
+                    hdr_offsets: info.offsets,
+                    meta: info.meta,
+                    flow,
+                    body_csum: info.body_csum,
+                    body,
+                    body_modified: false,
+                },
+            });
+        }
+
+        // Calculate the body offset within the new squashed segment
+        for s in &self.segs[..squash_to] {
+            body.seg_offset += s.len;
+        }
+        body.seg_index -= squash_to;
+
+        // Determine how big the message block for the squashed segment needs to
+        // be.
+        let mut new_seg_size = 0;
+        for s in &self.segs[..squash_to + 1] {
+            new_seg_size += s.len;
+        }
+
+        // Allocate a message block and copy in the squashed data.
+        let mut mp = allocb(new_seg_size);
+        unsafe {
+            for s in &self.segs[..squash_to + 1] {
+                core::ptr::copy_nonoverlapping(
+                    (*s.mp).b_rptr,
+                    (*mp).b_wptr,
+                    s.len,
+                );
+                (*mp).b_wptr = (*mp).b_wptr.add(s.len);
+            }
+        }
+
+        // Construct a new segment vector, tacking on any remaining segments
+        // after the header segments.
+        let orig_segs = core::mem::take(&mut self.segs);
+        let mut segs = vec![unsafe { PacketSeg::wrap_mblk(mp) }];
+        if squash_to + 1 < orig_segs.len() {
+            segs[0].link(&orig_segs[squash_to + 1]);
+            segs.extend_from_slice(&orig_segs[squash_to + 1..]);
+        }
+
         Ok(Packet {
             avail: self.avail,
             source: self.source,
-            // The new packet is taking ownership of the segments.
-            segs: core::mem::take(&mut self.segs),
+            segs,
             state: Parsed {
                 len: self.state.len,
                 hdr_offsets: info.offsets,
```

Comment:
All the header offsets in `info.offsets` need to be updated in this case.

Reply:
Done